twenty-first/src/shared_math/polynomial.rs

```diff
@@ -816,7 +560,270 @@ impl<PFElem: FiniteField> Polynomial<PFElem> {
+                }
+            }
+
+            // summand does not necessarily evaluate to 1 in domain[i],
+            // so we need to correct for this value
+            let mut summand_eval = zero;
+            for s in summand_array.iter().rev() {
+                summand_eval = summand_eval * domain[i] + *s;
+            }
+            let corrected_abscis = abscis / summand_eval;
+
+            // accumulate term
+            for j in 0..domain.len() {
+                lagrange_sum_array[j] += corrected_abscis * summand_array[j];
+            }
+        }
+        Polynomial {
+            coefficients: lagrange_sum_array,
+        }
+    }
+
+    pub fn are_colinear_3(
+        p0: (PFElem, PFElem),
+        p1: (PFElem, PFElem),
+        p2: (PFElem, PFElem),
+    ) -> bool {
+        if p0.0 == p1.0 || p1.0 == p2.0 || p2.0 == p0.0 {
+            return false;
+        }
+
+        let dy = p0.1 - p1.1;
+        let dx = p0.0 - p1.0;
+
+        dx * (p2.1 - p0.1) == dy * (p2.0 - p0.0)
+    }
+
+    pub fn get_colinear_y(p0: (PFElem, PFElem), p1: (PFElem, PFElem), p2_x: PFElem) -> PFElem {
+        debug_assert_ne!(p0.0, p1.0, "Line must not be parallel to y-axis");
+        let dy = p0.1 - p1.1;
+        let dx = p0.0 - p1.0;
+        let p2_y_times_dx = dy * (p2_x - p0.0) + dx * p0.1;
+
+        // Can we implement this without division?
+        p2_y_times_dx / dx
+    }
+
+    pub fn zerofier(domain: &[PFElem]) -> Self {
+        if domain.is_empty() {
+            return Self {
+                coefficients: vec![PFElem::one()],
+            };
+        }
+        let mut zerofier_array = vec![PFElem::zero(); domain.len() + 1];
+        zerofier_array[0] = PFElem::one();
+        let mut num_coeffs = 1;
+        for &d in domain.iter() {
+            for k in (1..num_coeffs + 1).rev() {
+                zerofier_array[k] = zerofier_array[k - 1] - d * zerofier_array[k];
+            }
+            zerofier_array[0] = -d * zerofier_array[0];
+            num_coeffs += 1;
+        }
+        Self {
+            coefficients: zerofier_array,
+        }
+    }
+
+    // Slow square implementation that does not use NTT
+    #[must_use]
+    pub fn slow_square(&self) -> Self {
+        let degree = self.degree();
+        if degree == -1 {
+            return Self::zero();
+        }
+
+        let squared_coefficient_len = self.degree() as usize * 2 + 1;
+        let zero = PFElem::zero();
+        let one = PFElem::one();
+        let two = one + one;
+        let mut squared_coefficients = vec![zero; squared_coefficient_len];
+
+        for i in 0..self.coefficients.len() {
+            let ci = self.coefficients[i];
+            squared_coefficients[2 * i] += ci * ci;
+
+            // TODO: Review.
+            for j in i + 1..self.coefficients.len() {
+                let cj = self.coefficients[j];
+                squared_coefficients[i + j] += two * ci * cj;
+            }
```

Comment:
- [x] Reviewed.

*(What are we assessing here, exactly?)*

Reply:
This is just code I moved around, since the MulAssign<BFieldElement> trait requirement had to be added to some methods.